### PR TITLE
Crash earlier for property type conflicts.

### PR DESCRIPTION
**Pull request by NightlyNexus**
IllegalArgumentException was the best exception type I could think of.

### CODE REVIEW

File: kotlin/src/test/java/com/squareup/moshi/KotlinJsonAdapterTest.kt [hunk 1]
```diff
@@ -296,6 +296,39 @@ class KotlinJsonAdapterTest {
...
+  }
+
+  @Test fun constructorParametersAndPropertiesWithSameNamesMustHaveSameTypes() {
+    val moshi = Moshi.Builder().add(KotlinJsonAdapterFactory()).build()
+    try {
+      moshi.adapter(ConstructorParameterWithSameNameAsPropertyButDifferentType::class.java)
+      fail()
+    } catch (expected: IllegalArgumentException) {
+      assertThat(expected).hasMessage(
+          "a has a constructor parameter of type kotlin.Int but property of type kotlin.String.")
```
Comment (JakeWharton) [inline kotlin/src/test/java/com/squareup/moshi/KotlinJsonAdapterTest.kt]: Probably worth single-quoting names. "a has a" reads funny.

File: kotlin/src/main/java/com/squareup/moshi/KotlinJsonAdapter.kt [hunk 2]
```diff
@@ -167,6 +167,12 @@ class KotlinJsonAdapterFactory : JsonAdapter.Factory {
       if (Modifier.isTransient(property.javaField?.modifiers ?: 0)) continue
 
       val parameter = parametersByName[property.name]
+
+      if (parameter != null && parameter.type != property.returnType) {
```
Comment (NightlyNexus) [inline kotlin/src/main/java/com/squareup/moshi/KotlinJsonAdapter.kt]: I think this is insufficient. We have to compare type qualifiers, too.

File: kotlin/src/test/java/com/squareup/moshi/KotlinJsonAdapterTest.kt [hunk 3]
```diff
@@ -296,6 +296,39 @@ class KotlinJsonAdapterTest {
...
+    val encoded = ConstructorParameterWithSameNameAsProperty(5)
+    assertThat(jsonAdapter.toJson(encoded)).isEqualTo("{\"a\":-1,\"b\":5}")
+
+    val decoded = jsonAdapter.fromJson("{\"a\":5}")!!
+    assertThat(decoded.a).isEqualTo(-1)
+    assertThat(decoded.b).isEqualTo(5)
+  }
+
+  class ConstructorParameterWithSameNameAsProperty(a: Int) {
+    var a = -1
```
Comment (NightlyNexus) [inline kotlin/src/test/java/com/squareup/moshi/KotlinJsonAdapterTest.kt]: I have a few concerns about this behavior that I want to check. Like what happens with type qualifiers or transient?
I'll look into it, ask questions, and follow up.

File: kotlin/src/main/java/com/squareup/moshi/KotlinJsonAdapter.kt (hunk 2 above)
Comment (NightlyNexus) [reply kotlin/src/main/java/com/squareup/moshi/KotlinJsonAdapter.kt]: no, never mind me.

**Comment by NightlyNexus** [comment]
tracking/noting other limitations in: #293

**Pull request merged**